$NetBSD$

Origin: EmberBSD (AI-assisted); upstream status: not submitted.

Select compiler/linker configuration without changing the host Perl ABI.

--- cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM.pm.orig
+++ cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM.pm
@@ -52,26 +52,26 @@
 }
 
 my %Is = ();
-$Is{VMS}    = $^O eq 'VMS';
-$Is{OS2}    = $^O eq 'os2';
-$Is{MacOS}  = $^O eq 'MacOS';
-if( $^O eq 'MSWin32' ) {
+$Is{VMS}    = $Config{osname} eq 'VMS';
+$Is{OS2}    = $Config{osname} eq 'os2';
+$Is{MacOS}  = $Config{osname} eq 'MacOS';
+if( $Config{osname} eq 'MSWin32' ) {
     _is_win95() ? $Is{Win95} = 1 : $Is{Win32} = 1;
 }
-$Is{UWIN}   = $^O =~ /^uwin(-nt)?$/;
-$Is{Cygwin} = $^O eq 'cygwin';
+$Is{UWIN}   = $Config{osname} =~ /^uwin(-nt)?$/;
+$Is{Cygwin} = $Config{osname} eq 'cygwin';
 $Is{NW5}    = $Config{osname} eq 'NetWare';  # intentional
-$Is{BeOS}   = ($^O =~ /beos/i or $^O eq 'haiku');
-$Is{DOS}    = $^O eq 'dos';
+$Is{BeOS}   = ($Config{osname} =~ /beos/i or $Config{osname} eq 'haiku');
+$Is{DOS}    = $Config{osname} eq 'dos';
 if( $Is{NW5} ) {
     $^O = 'NetWare';
     delete $Is{Win32};
 }
-$Is{VOS}    = $^O eq 'vos';
-$Is{QNX}    = $^O eq 'qnx';
-$Is{AIX}    = $^O eq 'aix';
-$Is{Darwin} = $^O eq 'darwin';
-$Is{OS390}  = $^O eq 'os390';
+$Is{VOS}    = $Config{osname} eq 'vos';
+$Is{QNX}    = $Config{osname} eq 'qnx';
+$Is{AIX}    = $Config{osname} eq 'aix';
+$Is{Darwin} = $Config{osname} eq 'darwin';
+$Is{OS390}  = $Config{osname} eq 'os390';
 
 $Is{Unix}   = !grep { $_ } values %Is;
 
--- cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Unix.pm.orig
+++ cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Unix.pm
@@ -30,7 +30,7 @@
     $Is{VMS}     = $^O eq 'VMS';
     $Is{OSF}     = $^O eq 'dec_osf';
     $Is{IRIX}    = $^O eq 'irix';
-    $Is{NetBSD}  = $^O eq 'netbsd';
+    $Is{NetBSD}  = $Config{osname} eq 'netbsd';
     $Is{Interix} = $^O eq 'interix';
     $Is{SunOS4}  = $^O eq 'sunos';
     $Is{Solaris} = $^O eq 'solaris';
@@ -39,7 +39,7 @@
                    grep( $^O eq $_, qw(bsdos interix dragonfly) )
                   );
     $Is{Android} = $^O =~ /android/;
-    if ( $^O eq 'darwin' ) {
+    if ( $Config{osname} eq 'darwin' ) {
       my @osvers = split /\./, $Config{osvers};
       if ( $^X eq '/usr/bin/perl' ) {
         $Is{ApplCor} = ( $osvers[0] >= 18 );
@@ -1084,7 +1084,7 @@
         if ( $Is{IRIX} ) {
             $ldrun = qq{-rpath "$self->{LD_RUN_PATH}"};
         }
-        elsif ( $^O eq 'darwin' && $Is{AppleRPath} ) {
+        elsif ( $Config{osname} eq 'darwin' && $Is{AppleRPath} ) {
             # both clang and gcc support -Wl,-rpath, but only clang supports
             # -rpath so by using -Wl,-rpath we avoid having to check for the
             # type of compiler
--- cpan/ExtUtils-MakeMaker/lib/ExtUtils/Liblist/Kid.pm.orig
+++ cpan/ExtUtils-MakeMaker/lib/ExtUtils/Liblist/Kid.pm
@@ -20,8 +20,8 @@
 use File::Spec;
 
 sub ext {
-    if    ( $^O eq 'VMS' )     { goto &_vms_ext; }
-    elsif ( $^O eq 'MSWin32' ) { goto &_win32_ext; }
+    if    ( $Config{osname} eq 'VMS' )     { goto &_vms_ext; }
+    elsif ( $Config{osname} eq 'MSWin32' ) { goto &_win32_ext; }
     else                       { goto &_unix_os2_ext; }
 }
 
@@ -49,7 +49,7 @@
     my ( $self, $potential_libs, $verbose, $give_libs ) = @_;
     $verbose ||= 0;
 
-    if ( $^O =~ /os2|android/ and $Config{perllibs} ) {
+    if ( $Config{osname} =~ /os2|android/ and $Config{perllibs} ) {
 
         # Dynamic libraries are not transitive, so we may need including
         # the libraries linked against perl.dll/libperl.so again.
@@ -84,7 +84,7 @@
     }
     @libpath = grep -d, @libpath;
 
-    if ($^O eq 'darwin')  {
+    if ($Config{osname} eq 'darwin')  {
         # 'escape' Mach-O ld -framework and -F flags, so they aren't dropped later on
         $found++ if $potential_libs =~ s/(^|\s)(-(?:weak_|reexport_|lazy_)?framework)\s+(\S+)/$1-Wl,$2 -Wl,$3/g;
         $found++ if $potential_libs =~ s/(^|\s)(-F)\s*(\S+)/$1-Wl,$2 -Wl,$3/g;
@@ -198,7 +198,7 @@
                 && -f ( $fullname = "$thispth/lib$thislib.$Config_dlext" ) )
             {
             }
-            elsif ( $^O eq 'darwin' && require DynaLoader && defined &DynaLoader::dl_load_file
+            elsif ( $Config{osname} eq 'darwin' && require DynaLoader && defined &DynaLoader::dl_load_file
                 && DynaLoader::dl_load_file( $fullname = "$thispth/lib$thislib.$so", 0 ) )
             {
             }
@@ -206,11 +206,11 @@
             }
             elsif ( -f ( $fullname = "$thispth/lib$thislib.dll$Config_libext" ) ) {
             }
-            elsif ( $^O eq 'cygwin' && -f ( $fullname = "$thispth/$thislib.dll" ) ) {
+            elsif ( $Config{osname} eq 'cygwin' && -f ( $fullname = "$thispth/$thislib.dll" ) ) {
             }
             elsif ( -f ( $fullname = "$thispth/Slib$thislib$Config_libext" ) ) {
             }
-            elsif ($^O eq 'dgux'
+            elsif ($Config{osname} eq 'dgux'
                 && -l ( $fullname = "$thispth/lib$thislib$Config_libext" )
                 && readlink( $fullname ) =~ /^elink:/s )
             {
@@ -239,7 +239,7 @@
 
             # what do we know about this library...
             # "Sounds like we should always assume it's a dynamic library on AIX."
-            my $is_dyna = $^O eq 'aix' ? 1 : ( $fullname !~ /\Q$Config_libext\E\z/ );
+            my $is_dyna = $Config{osname} eq 'aix' ? 1 : ( $fullname !~ /\Q$Config_libext\E\z/ );
             my $in_perl = ( $libs =~ /\B-l:?\Q${thislib}\E\b/s );
 
             # include the path to the lib once in the dynamic linker path
@@ -260,7 +260,7 @@
                 # For SunOS4, do not add in this shared library if
                 # it is already linked in the main perl executable
                 push( @ldloadlibs, "-l$custom_name$thislib" )
-                  unless ( $in_perl and $^O eq 'sunos' );
+                  unless ( $in_perl and $Config{osname} eq 'sunos' );
             }
             else {
                 push( @ldloadlibs, "-l$custom_name$thislib" );
--- cpan/ExtUtils-MakeMaker/lib/ExtUtils/MakeMaker/Config.pm.orig
+++ cpan/ExtUtils-MakeMaker/lib/ExtUtils/MakeMaker/Config.pm
@@ -10,6 +10,14 @@
 
 # Give us an overridable config.
 our %Config = %Config::Config;
+if (defined $ENV{EMBERBSD_PERL_BUILD_CONFIG}) {
+    require JSON::PP;
+    open my $fh, '<', $ENV{EMBERBSD_PERL_BUILD_CONFIG} or die $!;
+    local $/;
+    my $target = JSON::PP::decode_json(<$fh>);
+    %Config = %$target;
+}
+
 
 sub import {
     my $caller = caller;
--- dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder.pm.orig
+++ dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder.pm
@@ -4,6 +4,7 @@
 use File::Path ();
 use File::Basename ();
 use Perl::OSType qw/os_type/;
+use ExtUtils::MakeMaker::Config;
 
 use warnings;
 use strict;
@@ -22,10 +23,11 @@
 {
   my @package = split /::/, __PACKAGE__;
 
-  my $ostype = os_type();
-
-  if (grep {-e File::Spec->catfile($_, @package, 'Platform', $^O) . '.pm'} @INC) {
-      $load->(__PACKAGE__ . "::Platform::$^O");
+  my $target_os = $Config{osname};
+  my $ostype = os_type($target_os);
+
+  if (grep {-e File::Spec->catfile($_, @package, 'Platform', $target_os) . '.pm'} @INC) {
+      $load->(__PACKAGE__ . "::Platform::$target_os");
 
   } elsif ( $ostype &&
             grep {-e File::Spec->catfile($_, @package, 'Platform', $ostype) . '.pm'} @INC) {
--- dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder/Base.pm.orig
+++ dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder/Base.pm
@@ -4,7 +4,7 @@
 use File::Spec;
 use File::Basename;
 use Cwd ();
-use Config;
+use ExtUtils::MakeMaker::Config;
 use Text::ParseWords;
 use IPC::Cmd qw(can_run);
 use File::Temp qw(tempfile);
